contrib/dump-mtd-ifd.py

```diff
@@ -0,0 +1,185 @@
+    num_regions = (descriptor_map0 >> 24) & 0b111
+    if num_regions == 0:
+        num_regions = 10
+    print(f"num_regions=0x{num_regions:X}")
+
+    # read out FREGs
+    flash_region_base_addr = (descriptor_map0 >> 12) & 0x00000FF0
+    print(f"flash_region_base_addr=0x{flash_region_base_addr:X}")
+    flash_descriptor_regs: List[int] = []
+    for region in range(num_regions):
+        flash_descriptor_regs.append(
+            struct.unpack_from(
+                "<I", blob_ifd, offset=flash_region_base_addr + (region * 4)
+            )[0]
+        )
+    for region in range(num_regions):
+        print(f"flash_descriptor_reg{region}=0x{flash_descriptor_regs[region]:X}")
+
+    # parse each partition
+    fregs: List[IfdPartition] = []
+    for i in range(num_regions):
+        freg_base: int = (flash_descriptor_regs[i] << 12) & 0x07FFF000
+        freg_limit: int = ((flash_descriptor_regs[i] >> 4) & 0x07FFF000) | 0x00000FFF
+
+        # invalid
+        if freg_base > freg_limit:
+            continue
+
+        fregs.append(
+            IfdPartition(region=i, offset=freg_base, size=freg_limit - freg_base)
+        )
+
+    # create a binary blob big enough
+    image_size: int = 0
+    for freg in fregs:
+        if freg.offset + freg.size > image_size:
+            image_size = freg.offset + freg.size
+    print(f"image_size=0x{image_size:x}")
+    blob: bytearray = bytearray(image_size)
+
+    # copy each partition
+    for freg in fregs:
+        print("reading...", freg)
+        try:
+            f.seek(freg.offset)
+            blob[freg.offset : freg.offset + freg.size] = f.read(freg.size)
+        except OSError as e:
+            print(f"failed to read: {e}")
+    return blob
+
+
+def _read_device_to_file(devname: str, filename: Optional[str]) -> None:
+    # grab system info from sysfs
+    if not filename:
+        filename = ""
+        for sysfs_fn in [
+            "/sys/class/dmi/id/sys_vendor",
+            "/sys/class/dmi/id/product_family",
+            "/sys/class/dmi/id/product_name",
+            "/sys/class/dmi/id/product_sku",
+        ]:
+            try:
+                with open(sysfs_fn, "rb") as f:
+                    if filename:
+                        filename += "-"
+                    filename += (
+                        f.read().decode().replace("\n", "").replace(" ", "_").lower()
+                    )
+            except FileNotFoundError:
+                pass
+        if filename:
+            filename += ".bin"
+        else:
+            filename = "bios.bin"
+
+    # check this device name is what we expect
+    print(f"checking {devname}...")
+    try:
+        with open(f"/sys/class/mtd/{os.path.basename(devname)}/name", "rb") as f_name:
+            name = f_name.read().decode().replace("\n", "")
+    except FileNotFoundError as e:
+        sys.exit(str(e))
+    if name != "BIOS":
+        sys.exit(f"Not Intel Corporation PCH SPI Controller, got {name}")
```

Comment:
it's probably overkill as it adds a new dep; but you might look at `pyudev`.  I use it all the time in scripts that I need to get information about devices.

Reply:
Ya, overkill -- I need the dump script to run on stuff without a venv -- i.e. only stuff installed by default on RHEL/LTS etc.